File: src/Atlasd/Battlenet/Protocols/Game/Messages/SID_WARCRAFTGENERAL.cs
using Atlasd.Battlenet.Exceptions;
using Atlasd.Daemon;
using System;
using System.IO;
using System.Collections.Generic;

namespace Atlasd.Battlenet.Protocols.Game.Messages
{
    class SID_WARCRAFTGENERAL : Message
    {
        public enum SubCommands : byte
        {
            WID_GAMESEARCH = 0x00,
            WID_MAPLIST = 0x02,
            WID_CANCELSEARCH = 0x03,
            WID_USERRECORD = 0x04,
            WID_TOURNAMENT = 0x07,
            WID_CLANRECORD = 0x08,
            WID_ICONLIST = 0x09,
            WID_SETICON = 0x0A,
        };

        public SID_WARCRAFTGENERAL()
        {
            Id = (byte)MessageIds.SID_WARCRAFTGENERAL;
            Buffer = new byte[0];
        }

        public SID_WARCRAFTGENERAL(byte[] buffer)
        {
            Id = (byte)MessageIds.SID_WARCRAFTGENERAL;
            Buffer = buffer;
        }

        public override bool Invoke(MessageContext context)
        {
            Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)");

            if (Buffer.Length < 1)
                throw new GameProtocolViolationException(context.Client, $"{MessageName(Id)} buffer must be at least 1 byte");

            byte subcommand;
            using (var m = new MemoryStream(Buffer))
            using (var r = new BinaryReader(m))
                subcommand = r.ReadByte();

            Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} received subcommand {subcommand:X2}");

            // TODO: Compare subcommand variable with SubCommands enum and do procedures, for now just ignore the client

            return true;
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/Messages/SID_WARCRAFTGENERAL.vb
Imports AtlasV.Battlenet.Exceptions
Imports AtlasV.Daemon
Imports System
Imports System.IO
Imports System.Collections.Generic

Namespace AtlasV.Battlenet.Protocols.Game.Messages
    Class SID_WARCRAFTGENERAL
        Inherits Message

        Public Enum SubCommands As Byte
            WID_GAMESEARCH = &H0
            WID_MAPLIST = &H2
            WID_CANCELSEARCH = &H3
            WID_USERRECORD = &H4
            WID_TOURNAMENT = &H7
            WID_CLANRECORD = &H8
            WID_ICONLIST = &H9
            WID_SETICON = &HA
        End Enum

        Public Sub New()
            Id = CByte(MessageIds.SID_WARCRAFTGENERAL)
            Buffer = Array.Empty(Of Byte)()
        End Sub

        Public Sub New(ByVal varBuffer As Byte())
            Id = CByte(MessageIds.SID_WARCRAFTGENERAL)
            Buffer = varBuffer
        End Sub

        Public Overrides Function Invoke(ByVal context As MessageContext) As Boolean
            Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")
            If context.Client.GameState Is Nothing OrElse Not Product.IsWarcraftIII(context.Client.GameState.Product) Then Throw New GameProtocolViolationException(context.Client, $"{MessageName(Id)} is Warcraft III game client exclusive")
            If Buffer.Length < 1 Then Throw New GameProtocolViolationException(context.Client, $"{MessageName(Id)} buffer must be at least 1 byte")
            Dim subcommand As Byte

            Using m = New MemoryStream(Buffer)

                Using r = New BinaryReader(m)
                    subcommand = r.ReadByte()
                End Using
            End Using

            Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} received subcommand {subcommand}")
            Return True
        End Function
    End Class
End Namespace
